File: nodejs/3.1/pubnub.js
/* ---------------------------------------------------------------------------
WAIT! - This file depends on instructions from the PUBNUB Cloud.
http://www.pubnub.com/account
--------------------------------------------------------------------------- */

/* ---------------------------------------------------------------------------
PubNub Real-time Cloud-Hosted Push API and Push Notification Client Frameworks
Copyright (c) 2011 TopMambo Inc.
http://www.pubnub.com/
http://www.pubnub.com/terms
--------------------------------------------------------------------------- */

/* ---------------------------------------------------------------------------
Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.
--------------------------------------------------------------------------- */

/**
 * UTIL LOCALS
 */
var NOW    = 1
,   http   = require('http')
,   https  = require('https')
,   URLBIT = '/'
,   XHRTME = 310000
,   XORIGN = 1;

/**
 * UNIQUE
 * ======
 * var timestamp = unique();
 */
function unique() { return'x'+ ++NOW+''+(+new Date) }


/**
 * NEXTORIGIN
 * ==========
 * var next_origin = nextorigin();
 */
var nextorigin = (function() {
    var ori = Math.floor(Math.random() * 9) + 1;
    return function(origin) {
        return origin.indexOf('pubsub') > 0
            && origin.replace(
             'pubsub', 'ps' + (++ori < 10 ? ori : ori=1)
            ) || origin;
    }
})();

/**
 * LOG
 * ===
 * log('message');
 */
function log(message) { console['log'](message) }

/**
 * timeout
 * =======
 * timeout( function(){}, 100 );
 */
function timeout( fun, wait ) { return setTimeout( fun, wait ) }

/**
 * ENCODE
 * ======
 * var encoded_path = encode('path');
 */
function encode(path) {
    return (''+path).split('').map(function(chr) {
        return " ~`!@#$%^&*()+=[]\\{}|;':\",./<>?".indexOf(chr) > -1 ?
               "%"+chr.charCodeAt(0).toString(16).toUpperCase()      : chr
    }).join('');
}

/**
 * Request
 * =======
 *  xdr({
 *     url     : ['http://www.blah.com/url'],
 *     success : function(response) {},
 *     fail    : function() {}
 *  });
 */
function xdr( setup ) {
    setup.url.unshift('');

    var url     = setup.url.join(URLBIT)
    ,   success = setup.success
    ,   origin  = setup.origin
    ,   ssl     = setup.ssl
    ,   failed  = 0
    ,   body    = ''
    ,   fail    = function(e) {
            if (failed) return;
            failed = 1;
            (setup.fail||function(){})(e);
        };

    try {
        (ssl ? https : http).get( {
            host : origin,
            port : ssl ? 443 : 80,
            path : url
        }, function(response) {
            response.setEncoding('utf8');
            response.on( 'error', fail );
            response.on( 'data', function (chunk) {
                if (chunk) body += chunk;
            } );
            response.on( 'end', function () {
                try {
                    if (body) return success(JSON.parse(body));
                    else      fail();
                } catch(e) { fail(); }
            } );
        }).on( 'error', fail );
    } catch(e) { fail(); }
}



/* =-=====================================================================-= */
/* =-=====================================================================-= */
/* =-=========================     PUBNUB     ============================-= */
/* =-=====================================================================-= */
/* =-=====================================================================-= */

exports.init = function(setup) {
    var DEMO          = 'demo'
    ,   PUBLISH_KEY   = setup.publish_key   || DEMO
    ,   SUBSCRIBE_KEY = setup.subscribe_key || DEMO
    ,   SECRET_KEY    = setup.secret_key    || false
    ,   SSL           = setup.ssl
    ,   ORIGIN        = setup.origin || 'pubsub.pubnub.com'
    ,   CHANNELS      = {}
    ,   PN            = {
        /*
            PUBNUB.history({
                channel  : 'my_chat_channel',
                limit    : 100,
                callback : function(messages) { console.log(messages) }
            });
        */
        history : function( args, callback ) {
            var callback = args['callback']
            ,   limit    = args['limit'] || 10
            ,   channel  = args['channel'];

            // Make sure we have a Channel
            if (!channel)  return log('Missing Channel');
            if (!callback) return log('Missing Callback');

            // Send Message
            xdr({
                ssl : SSL,
                url : [
                    'history',
                    SUBSCRIBE_KEY, encode(channel),
                    '0', limit
                ],
                origin  : ORIGIN,
                success : callback,
                fail    : function(response) { log(response) }
            });
        },

        /*
            PUBNUB.time(function(time){ console.log(time) });
        */
        time : function(callback) {
            xdr({
                ssl     : SSL,
                url     : ['time', '0'],
                origin  : ORIGIN,
                success : function(response) { callback(response[0]) },
                fail    : function() { callback(0) }
            });
        },

        /*
            PUBNUB.publish({
                channel : 'my_chat_channel',
                message : 'hello!'
            });
        */
        publish : function( args, callback ) {
            var callback = callback || args['callback'] || function(){}
            ,   message  = args['message']
            ,   channel  = args['channel'];

            if (!message)     return log('Missing Message');
            if (!channel)     return log('Missing Channel');
            if (!PUBLISH_KEY) return log('Missing Publish Key');

            // If trying to send Object
            message = JSON['stringify'](message);

            // Send Message
            xdr({
                ssl      : SSL,
                origin   : ORIGIN,
                success  : function(response) { callback(response) },
                fail     : function() { callback([ 0, 'Disconnected' ]) },
                url      : [
                    'publish',
                    PUBLISH_KEY, SUBSCRIBE_KEY,
                    0, encode(channel),
                    '0', encode(message)
                ]
            });
        },

        /*
            PUBNUB.unsubscribe({ channel : 'my_chat' });
        */
        unsubscribe : function(args) {
            var channel = args['channel'];

            // Leave if there never was a channel.
            if (!(channel in CHANNELS)) return;

            // Disable Channel
            CHANNELS[channel].connected = 0;

            // Abort and Remove Script
            CHANNELS[channel].done && 
            CHANNELS[channel].done(0);
        },

        /*
            PUBNUB.subscribe({
                channel  : 'my_chat'
                callback : function(message) { console.log(message) }
            });
        */
        'subscribe' : function( args, callback ) {

            var channel      = args['channel']
            ,   callback     = callback || args['callback']
            ,   timetoken    = 0
            ,   error        = args['error'] || function(){}
            ,   connect      = args['connect'] || function(){}
            ,   reconnect    = args['reconnect'] || function(){}
            ,   disconnect   = args['disconnect'] || function(){}
            ,   disconnected = 0
            ,   connected    = 0
            ,   origin       = nextorigin(ORIGIN);

            // Make sure we have a Channel
            if (!channel)       return log('Missing Channel');
            if (!callback)      return log('Missing Callback');
            if (!SUBSCRIBE_KEY) return log('Missing Subscribe Key');

            if (!(channel in CHANNELS)) CHANNELS[channel] = {};

            // Make sure we have a Channel
            if (CHANNELS[channel].connected) return log('Already Connected');
                CHANNELS[channel].connected = 1;

            // Recurse Subscribe
            function pubnub() {
                // Stop Connection
                if (!CHANNELS[channel].connected) return;

                // Connect to PubNub Subscribe Servers
                CHANNELS[channel].done = xdr({
                    ssl    : SSL,
                    origin : origin,
                    url    : [
                        'subscribe',
                        SUBSCRIBE_KEY,
                        encode(channel),
                        '0', timetoken
                    ],
                    fail : function() {
                        // Disconnect
                        if (!disconnected) {
                            disconnected = 1;
                            disconnect();
                        }
                        timeout( pubnub, 1000 );
                        PN['time'](function(success){
                            success || error();
                        });
                    },
                    success : function(messages) {
                        if (!CHANNELS[channel].connected) return;

                        // Connect
                        if (!connected) {
                            connected = 1;
                            connect();
                        }

                        // Reconnect
                        if (disconnected) {
                            disconnected = 0;
                            reconnect();
                        }

                        messages[0].forEach(function(msg) {
                            callback( msg, messages );
                        });

                        timetoken = messages[1];
                        timeout( pubnub, 10 );
                    }
                });
            }

            // Begin Recursive Subscribe
            pubnub();
        },
    };

    return PN;
}

File: csharp/3.1/Pubnub.cs
using System;
using System.IO;
using System.Net;
using System.Text;
using System.Collections;
using System.Collections.Generic;
using System.Security.Cryptography;
using System.Web.Script.Serialization;

/**
 * PubNub 3.0 Real-time Push Cloud API
 *
 * @author Stephen Blum
 * @package pubnub
 */
public class PubnubTEST {
    static public void Main() {
        // -----------------
        // Init Pubnub Class
        // -----------------
        Pubnub pubnub = new Pubnub(
            "demo",  // PUBLISH_KEY
            "demo",  // SUBSCRIBE_KEY
            "",      // SECRET_KEY
            false    // SSL_ON?
        );
        string channel = "test-channel-顶";

        // ---------------
        // Publish Message
        // ---------------
        List<object> info = pubnub.Publish( channel, "Hello World" );

        // ----------------
        // Publish Response
        // ----------------
        Console.WriteLine(
            "Publish Success: " + info[0].ToString() +
            "\nPublish Info: "  + info[1]
        );

        // -------
        // History
        // -------
        List<object> history = pubnub.History( channel, 1 );
        foreach (object history_message in history) {
            Console.Write("History Message: ");
            Console.WriteLine(history_message);
        }

        // ----------------------
        // Get PubNub Server Time
        // ----------------------
        object timestamp = pubnub.Time();
        Console.WriteLine("Server Time: " + timestamp.ToString());

        // ---------
        // Subscribe
        // ---------
        pubnub.Subscribe(
            channel,
            delegate (object message) {
                Console.WriteLine("Received Message -> '" + message + "'");
                return true;
            }
        );
    }
}


/**
 * PubNub 3.0 Real-time Push Cloud API
 *
 * @author Stephen Blum
 * @package pubnub
 */
public class Pubnub {
    private string ORIGIN        = "pubsub.pubnub.com";
    private int    LIMIT         = 1800;
    private string PUBLISH_KEY   = "";
    private string SUBSCRIBE_KEY = "";
    private string SECRET_KEY    = "";
    private bool   SSL           = false;

    public delegate bool Procedure(object message);

    /**
     * PubNub 3.0
     *
     * Prepare PubNub Class State.
     *
     * @param string Publish Key.
     * @param string Subscribe Key.
     * @param string Secret Key.
     * @param bool SSL Enabled.
     */
    public Pubnub(
        string publish_key,
        string subscribe_key,
        string secret_key,
        bool ssl_on
    ) {
        this.init( publish_key, subscribe_key, secret_key, ssl_on );
    }

    /**
     * PubNub 2.0 Compatibility
     *
     * Prepare PubNub Class State.
     *
     * @param string Publish Key.
     * @param string Subscribe Key.
     */
    public Pubnub(
        string publish_key,
        string subscribe_key
    ) {
        this.init( publish_key, subscribe_key, "", false );
    }

    /**
     * PubNub 3.0 without SSL
     *
     * Prepare PubNub Class State.
     *
     * @param string Publish Key.
     * @param string Subscribe Key.
     * @param string Secret Key.
     */
    public Pubnub(
        string publish_key,
        string subscribe_key,
        string secret_key
    ) {
        this.init( publish_key, subscribe_key, secret_key, false );
    }

    /**
     * Init
     *
     * Prepare PubNub Class State.
     *
     * @param string Publish Key.
     * @param string Subscribe Key.
     * @param string Secret Key.
     * @param bool SSL Enabled.
     */
    public void init(
        string publish_key,
        string subscribe_key,
        string secret_key,
        bool ssl_on
    ) {
        this.PUBLISH_KEY   = publish_key;
        this.SUBSCRIBE_KEY = subscribe_key;
        this.SECRET_KEY    = secret_key;
        this.SSL           = ssl_on;

        // SSL On?
        if (this.SSL) {
            this.ORIGIN = "https://" + this.ORIGIN;
        }
        else {
            this.ORIGIN = "http://" + this.ORIGIN;
        }
    }

    /**
     * History
     *
     * Load history from a channel.
     *
     * @param String channel name.
     * @param int limit history count response.
     * @return ListArray of history.
     */
    public List<object> History( string channel, int limit ) {
        List<string> url = new List<string>();

        url.Add("history");
        url.Add(this.SUBSCRIBE_KEY);
        url.Add(channel);
        url.Add("0");
        url.Add(limit.ToString());

        return _request(url);
    }

    /**
     * Publish
     *
     * Send a message to a channel.
     *
     * @param String channel name.
     * @param List<object> info.
     * @return bool false on fail.
     */
    public List<object> Publish( string channel, object message ) {
        JavaScriptSerializer serializer = new JavaScriptSerializer();

        // Generate String to Sign
        string signature = "0";
        if (this.SECRET_KEY.Length > 0) {
            StringBuilder string_to_sign = new StringBuilder();
            string_to_sign
                .Append(this.PUBLISH_KEY)
                .Append('/')
                .Append(this.SUBSCRIBE_KEY)
                .Append('/')
                .Append(this.SECRET_KEY)
                .Append('/')
                .Append(channel)
                .Append('/')
                .Append(serializer.Serialize(message));

            // Sign Message
            signature = md5(string_to_sign.ToString());
        }

        // Build URL
        List<string> url = new List<string>();
        url.Add("publish");
        url.Add(this.PUBLISH_KEY);
        url.Add(this.SUBSCRIBE_KEY);
        url.Add(signature);
        url.Add(channel);
        url.Add("0");
        url.Add(serializer.Serialize(message));

        // Return JSONArray
        return _request(url);
    }

    /**
     * Subscribe
     *
     * This function is BLOCKING.
     * Listen for a message on a channel.
     *
     * @param string channel name.
     * @param Procedure function callback.
     */
    public void Subscribe( string channel, Procedure callback ) {
        this._subscribe( channel, callback, 0 );
    }

    /**
     * Subscribe - Private Interface
     *
     * @param string channel name.
     * @param Procedure function callback.
     * @param string timetoken.
     */
    private void _subscribe(
        string    channel,
        Procedure callback,
        object    timetoken
    ) {
        // Begin Recusive Subscribe
        try {
            // Build URL
            List<string> url = new List<string>();
            url.Add("subscribe");
            url.Add(this.SUBSCRIBE_KEY);
            url.Add(channel);
            url.Add("0");
            url.Add(timetoken.ToString());

            // Wait for Message
            List<object> response = _request(url);

            // Update TimeToken
            if (response[1].ToString().Length  > 0)
                timetoken = (object)response[1];

            // Run user Callback and Reconnect if user permits.
            foreach (object message in (object[])response[0]) {
                if (!callback(message)) return;
            }

            // Keep listening if Okay.
            this._subscribe( channel, callback, timetoken );
        }
        catch {
            System.Threading.Thread.Sleep(1000);
            this._subscribe( channel, callback, timetoken );
        }
    }

    /**
     * Time
     *
     * Timestamp from PubNub Cloud.
     *
     * @return object timestamp.
     */
    public object Time() {
        List<string> url = new List<string>();

        url.Add("time");
        url.Add("0");

        List<object> response = _request(url);
        return response[0];
    }

    /**
     * Request URL
     *
     * @param List<string> request of url directories.
     * @return List<object> from JSON response.
     */
    private List<object> _request(List<string> url_components) {
        string        temp  = null;
        int           count = 0;
        byte[]        buf   = new byte[8192];
        StringBuilder url   = new StringBuilder();
        StringBuilder sb    = new StringBuilder();

        JavaScriptSerializer serializer = new JavaScriptSerializer();

        // Add Origin To The Request
        url.Append(this.ORIGIN);

        // Generate URL with UTF-8 Encoding
        foreach ( string url_bit in url_components) {
            url.Append("/");
            url.Append(_encodeURIcomponent(url_bit));
        }

        // Fail if string too long
        if (url.Length > this.LIMIT) {
            List<object> too_long = new List<object>();
            too_long.Add(0);
            too_long.Add("Message Too Long.");
            return too_long;
        }

        // Create Request
        HttpWebRequest  request   = (HttpWebRequest)
            WebRequest.Create(url.ToString());

        // Set Timeout
        request.Timeout          = 200000;
        request.ReadWriteTimeout = 200000;

        // Receive Response
        HttpWebResponse response  = (HttpWebResponse)request.GetResponse();
        Stream          resStream = response.GetResponseStream();

        // Read
        do {
            count = resStream.Read( buf, 0, buf.Length );
            if (count != 0) {
                temp = Encoding.UTF8.GetString( buf, 0, count );
                sb.Append(temp);
            }
        } while (count > 0);

        // Parse Response
        string message = sb.ToString();

        return serializer.Deserialize<List<object>>(message);
    }

    private string _encodeURIcomponent(string s) {
        StringBuilder o = new StringBuilder();
        foreach (char ch in s.ToCharArray()) {
            if (isUnsafe(ch)) {
                o.Append('%');
                o.Append(toHex(ch / 16));
                o.Append(toHex(ch % 16));
            }
            else o.Append(ch);
        }
        return o.ToString();
    }

    private char toHex(int ch) {
        return (char)(ch < 10 ? '0' + ch : 'A' + ch - 10);
    }

    private bool isUnsafe(char ch) {
        return " ~`!@#$%^&*()+=[]\\{}|;':\",./<>?".IndexOf(ch) >= 0;
    }

    private static string md5(string text) {
        MD5 md5 = new MD5CryptoServiceProvider();
        byte[] data = Encoding.Default.GetBytes(text);
        byte[] hash = md5.ComputeHash(data);
        string hexaHash = "";
        foreach (byte b in hash) hexaHash += String.Format("{0:x2}", b);
        return hexaHash;
    }
}
 
